docs/proposals/pod-resource-management.md

```diff
@@ -0,0 +1,394 @@
+
+## Implementation Plan
+
+The implementation plan is outlined in the next sections.
+We will have a 'enable-qos-cgroups' flag to specify if the user wants to use the QoS based cgroup hierarchy. The flag would be set to false by default at least in v1.4.
+
+#### Top level Cgroups for QoS tiers
+
+Two top level cgroups for Burstable and Best effort Qos classes are created when Kubelet starts to run on a node. All guaranteed pods cgroups are by default nested under the `Root`. So we dont create a top level cgroup for the Guaranteed class. For raw cgroup systems we would use libcontainers cgroups manager for general cgroup management(cgroup creation/destruction). But for systemd we don't have equivalent support for slice management in libcontainer yet. So we will be adding support for the same in the Kubelet. These cgroups are only created once on Kubelet initialization as a part of node setup.
+
+#### Pod level Cgroup creation and deletion (Docker runtime)
+
+- When a new pod is brought up, its QoS class is firstly determined.
+- We add an interface to Kubelet’s ContainerManager to create and delete pod level cgroups under the cgroup that matches the pod’s QoS class.
+- This interface will be pluggable. Kubelet will support both systemd and raw cgroups based __cgroup__ drivers. We will be using the --cgroup-driver flag proposed in the [Systemd Node Spec](kubelet-systemd.md) to specify the cgroup driver.
+- We inject creation and deletion of pod level cgroups into the pod workers.
+- As new pods are added QoS class cgroup parameters are updated to match the resource requests by the Pods.
+- To ensure Kubelet's idempotent behaviour we follow a pod cgroup naming format which is opaque and deterministic. We will use the pod UID as the name of the cgroups.
+
+#### Container level cgroups
+
+Have docker manager create container cgroups under pod level cgroups. For determining the parent pod cgroup name and path we have a separate implementation for systemd and raw cgroup systems. This can be easily implemented in raw cgroup systems as now docker has support for specifying parent flag of cgroups. In systemd systems we would have to add support for specifying the parent slice during scope creation.
+
+#### Rkt runtime
+
+We want to have rkt create pods under a root QoS class that kubelet specifies, and set pod level cgroup parameters mentioned in this proposal by itself.
+
+#### Add Pod level metrics to Kubelet's metrics provider
+
+Update Kubelet’s metrics provider to include Pod level metrics. Use cAdvisor's cgroup subsystem information to determine various Pod level usage metrics.
```

Comment:
Will we update the `kubelet` summary stats provider to return pod level usage as well.  Right now, its just containers.